Detectors/MUON/MID/Raw/include/MIDRaw/ELinkManager.h

```diff
@@ -0,0 +1,80 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \file   MIDRaw/ELinkManager.h
+/// \brief  MID e-link data shaper manager
+/// \author Diego Stocco <Diego.Stocco at cern.ch>
+/// \date   18 March 2021
+#ifndef O2_MID_ELINKMANAGER_H
+#define O2_MID_ELINKMANAGER_H
+
+#include <cstdint>
+#include <vector>
+// #include <unordered_map>
+#include "MIDRaw/ELinkDataShaper.h"
+#include "MIDRaw/ELinkDecoder.h"
+#include "MIDRaw/ElectronicsDelay.h"
+#include "MIDRaw/FEEIdConfig.h"
+
+namespace o2
+{
+namespace mid
+{
+class ELinkManager
+{
+ public:
+  void init(uint16_t feeId, bool isDebugMode, bool isBare = false, const ElectronicsDelay& electronicsDelay = ElectronicsDelay(), const FEEIdConfig& feeIdConfig = FEEIdConfig());
+
+  void set(uint32_t orbit);
+
+  /// Main function to be executed when decoding is done
+  inline void onDone(const ELinkDecoder& decoder, uint8_t boardUniqueId, std::vector<ROBoard>& data, std::vector<ROFRecord>& rofs) { return onDone(decoder, raw::getCrateId(boardUniqueId), raw::getLocId(boardUniqueId), data, rofs); }
+
+  /// Main function to be executed when decoding is done
+  inline void onDone(const ELinkDecoder& decoder, std::vector<ROBoard>& data, std::vector<ROFRecord>& rofs) { return onDone(decoder, decoder.getCrateId(), decoder.getId(), data, rofs); }
+
+  // Use vectors
+
```

Comment:
This was actually done on purpose, since I developed the two ways and I wanted to keep track of it. The vector implementation is faster, but less easily configurable, while the unordered_map implementation is more flexible, but slower than the vector one. I don't plan to use them at the same time, but I might want to switch at some point. So, not enough motivation to have two separate and configurable implementations.
Are preprocessor directives allowed in the project?

Reply:
I believe so. At least I've used them :wink: (see e.g. https://github.com/AliceO2Group/AliceO2/blob/dev/Detectors/MUON/Common/src/dcs-ccdb.cxx and https://github.com/AliceO2Group/AliceO2/blob/dev/Detectors/MUON/Common/CMakeLists.txt)

Reply:
Ok, I added preprocessor directives